go/ephemeral/user_ek_box_storage.go

```diff
@@ -0,0 +1,249 @@
+	Result struct {
+		Box                string                `json:"box"`
+		DeviceEKGeneration keybase1.EkGeneration `json:"device_ek_generation"`
+		Sig                string                `json:"sig"`
+	} `json:"result"`
+}
+
+func (s *UserEKBoxStorage) fetchAndPut(ctx context.Context, generation keybase1.EkGeneration) (userEK keybase1.UserEk, err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#fetchAndPut", func() error { return err })()
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/user_ek_box",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"generation":          libkb.U{Val: uint64(generation)},
+			"recipient_device_id": libkb.S{Val: string(s.G().Env.GetDeviceID())},
+		},
+	}
+
+	var result UserEKBoxedResponse
+	res, err := s.G().GetAPI().Get(apiArg)
+	if err != nil {
+		return userEK, err
+	}
+
+	err = res.Body.UnmarshalAgain(&result)
+	if err != nil {
+		return userEK, err
+	}
+
+	// Before we store anything, let's verify that the server returned
+	// signature is valid and the KID it has signed matches the boxed seed.
+	// Otherwise something's fishy..
+	userEKMetadata, wrongKID, err := VerifySigWithLatestPUK(ctx, s.G(), result.Result.Sig)
+
+	// Check the wrongKID condition before checking the error, since an error
+	// is still returned in this case. TODO: Turn this warning into an error
+	// after EK support is sufficiently widespread.
+	if wrongKID {
+		s.G().Log.CWarningf(ctx, "It looks like you revoked a device without generating new ephemeral keys. Are you running an old version?")
+		return userEK, nil
+	}
+	if err != nil {
+		return userEK, err
+	}
+
+	if userEKMetadata == nil { // shouldn't happen
+		s.G().Log.CWarningf(ctx, "No error but got nil userEKMetadata")
+		return userEK, err
+	}
+
+	userEKBoxed := keybase1.UserEkBoxed{
+		Box:                result.Result.Box,
+		DeviceEkGeneration: result.Result.DeviceEKGeneration,
+		Metadata:           *userEKMetadata,
+	}
+
+	userEK, err = s.unbox(ctx, userEKBoxed)
+	if err != nil {
+		return userEK, err
+	}
+
+	seed := UserEKSeed(userEK.Seed)
+	keypair, err := seed.DeriveDHKey()
+	if err != nil {
+		return userEK, err
+
+	}
+
+	if !keypair.GetKID().Equal(userEKMetadata.Kid) {
+		return userEK, fmt.Errorf("Failed to verify server given seed against signed KID %s", userEKMetadata.Kid)
+	}
+
+	// Store the boxed version, return the unboxed
+	err = s.put(generation, userEKBoxed)
+	return userEK, err
+}
+
+func (s *UserEKBoxStorage) Put(ctx context.Context, generation keybase1.EkGeneration, userEKBoxed keybase1.UserEkBoxed) (err error) {
+	defer s.G().CTrace(ctx, "UserEKBoxStorage#Put", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+	return s.put(generation, userEKBoxed)
+}
+
+func (s *UserEKBoxStorage) put(generation keybase1.EkGeneration, userEKBoxed keybase1.UserEkBoxed) (err error) {
+	key := s.dbKey()
+	s.cache[generation] = userEKBoxed
+	err = s.G().GetKVStore().PutObj(key, nil, s.cache)
```

Comment:
Could this be a bug if `index` hasn't been called? Is it possible we'd stomp on generations that are present on disk but not in memory?

It might be safer to refactor `index()` into a `getCache()` method, that lazily indexes the first time you call it. That way most callers can't really make the mistake of forgetting it.

Reply:
yeah i liek that a lot more